skimage/feature/_sift.pyx

```diff
@@ -0,0 +1,167 @@
+        histogram bin.  Shape (n_hist, n_patch).
+    rc_bin_spacing : float
+        The spacing between spatial histogram bins along a single axis.
+
+    """
+    cdef:
+        Py_ssize_t i, p, r, c, k_index, k_index2
+        Py_ssize_t n_patch = len(magnitude)
+        Py_ssize_t n_hist = histograms.shape[0]
+        Py_ssize_t n_ori = histograms.shape[2]
+        np_floats t_val, val_norm1, w0, w1, w2, inv_spacing
+    inv_spacing = 1 / rc_bin_spacing
+    val_norm1 = n_ori / (2 * M_PI)
+    for p in range(n_patch):
+        for r in range(n_hist):
+            if dist_r[r, p] > rc_bin_spacing:
+                continue
+            for c in range(n_hist):
+                if dist_c[c, p] > rc_bin_spacing:
+                    continue
+                w0 = ((1 - inv_spacing * dist_r[r, p])
+                      * (1 - inv_spacing * dist_c[c, p])
+                      * magnitude[p])
+                t_val = near_t_val[p]
+                w1 = w0 * val_norm1 * t_val
+                w2 = w0 - w1
+                k_index = near_t[p]
+                if k_index == n_ori - 1:
+                    k_index2 = 0
+                else:
+                    k_index2 = k_index + 1
+                histograms[r, c, k_index] += w1
+                histograms[r, c, k_index2] += w2
+
+
+cpdef _local_max(np_floats[:, :, ::1] octave, double thresh):
+    cdef:
+        Py_ssize_t n_r = octave.shape[0]
+        Py_ssize_t n_c = octave.shape[1]
+        Py_ssize_t n_s = octave.shape[2]
+        Py_ssize_t r, c, s
+        np_floats* center_ptr
+        Py_ssize_t neighbor_offsets[26]
+        np_floats center_val, val
+        int n = 0
+
+    for r in range(-1, 2):
+        for c in range(-1, 2):
+            for s in range(-1, 2):
+                if (r != 0 or c != 0 or s != 0):
+                    neighbor_offsets[n] = (r * n_c + c) * n_s + s
+                    n += 1
+
+    maxima_coords = []
+    for r in range(1, n_r - 1):
+        for c in range(1, n_c - 1):
+            for s in range(1, n_s - 1):
+                center_ptr = &octave[r, c, s]
+                center_val = center_ptr[0]
+                if abs(center_val) < thresh:
+                    continue
+                is_local_min = True
+                for offset in neighbor_offsets:
+                    val = center_ptr[offset]
+                    if val <= center_val:
+                        is_local_min = False
+                        break
+
+                if is_local_min:
+                    is_local_max = False
+                else:
+                    is_local_max = True
+                    for offset in neighbor_offsets:
+                        val = center_ptr[offset]
+                        if val >= center_val:
+                            is_local_max = False
+                            break
+                if is_local_min or is_local_max:
+                    maxima_coords.append((r, c, s))
```

Comment:
I think that we can simplify this block:
```suggestion
                for offset in neighbor_offsets:
                    val = center_ptr[offset]
                    if val <= center_val:
                        for offset in neighbor_offsets:
                            val = center_ptr[offset]
                            if val >= center_val:
                                break
                        else:
                            # It is a local max
                            maxima_coords.append((r, c, s))
                    break
                else:
                     # It is a local min
                    maxima_coords.append((r, c, s))
```

Reply:
Right now, this change doesn't work, at least the tests are failing

Reply:
Well, I am not sure to understand why :confused:. This refactoring should be equivalent
```suggestion
                is_local_min = True
                is_local_max = False
                for offset in neighbor_offsets:
                    val = center_ptr[offset]
                    if val <= center_val:
                        is_local_max = True
                        is_local_min = False
                        for offset in neighbor_offsets:
                            val = center_ptr[offset]
                            if val >= center_val:
                                is_local_max = False
                                break
                        break
                if is_local_min or is_local_max:
                    maxima_coords.append((r, c, s))
```

Reply:
@faberno

> Right now, this change doesn't work, at least the tests are failing

just to be sure, the `else` statements must be aligned with the `for` bloc, not the `if` bloc ^^

Reply:
The second version would work. I'm a bit tired so correct me if I'm wrong: Aren't you checking the neighbors unnecessarily often by nesting the loops?
I guess we could put everything in one loop if we wanted to, but to me the initial solution looks quite good and intuitive :D

Reply:
> Aren't you checking the neighbors unnecessarily often by nesting the loops?

I don't think so, the `break` statements stops the search whenever it is done...

> I guess we could put everything in one loop if we wanted to, but to me the initial solution looks quite good and intuitive :D

By putting everything in one loop, I just want to avoid an `if ... else ...` bloc when looping over all the elements of the `octave` array...

Reply:
You're right, your version is a bit faster, I will add it :D

Reply:
Ahh, you're first version didn't work because the second break was wrongly indented. But I tested it and the second version with the booleans seems to be a tiny bit faster

Reply:
OK, thank you @faberno for your invetigations :wink: